File: examples/custom_operator.f90
!
!  This file is part of HPDDM.

!  Author(s): Pierre Jolivet <pierre.jolivet@enseeiht.fr>
!       Date: 2016-05-01

!  Copyright (C) 2016-     Centre National de la Recherche Scientifique

!  HPDDM is free software: you can redistribute it and/or modify
!  it under the terms of the GNU Lesser General Public License as published
!  by the Free Software Foundation, either version 3 of the License, or
!  (at your option) any later version.

!  HPDDM is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!  GNU Lesser General Public License for more details.

!  You should have received a copy of the GNU Lesser General Public License
!  along with HPDDM.  If not, see <http://www.gnu.org/licenses/>.
!

      module my_operator
          implicit none
          contains
              subroutine my_mv(n, in, out, mu)
                  use, intrinsic :: iso_c_binding, only: c_int, c_double
                  use mpi
                  ! size of the local matrix
                  integer (c_int), intent (in) :: n
                  ! number of right-hand sides
                  integer (c_int), intent (in) :: mu
                  ! input vector x
                  real (c_double), intent (in), dimension(n, mu)        &
     &                                                       :: in
                  ! output vector y = A x
                  real (c_double), intent (inout), dimension(n, mu)     &
     &                                                       :: out
                  integer :: i, j, rank
                  call MPI_Comm_rank(MPI_COMM_WORLD, rank, i)
                  ! simple tridiagonal system
                  ! no exchange between processes because the first
                  ! and the last off-diagonal coefficients are removed
                  do 20 j = 1, mu
                      do 30 i = 1, n
                          if(i > 1 .and. i < n) then
                              out(i, j) = (n * rank + i) * in(i, j)     &
     &                                  + 0.5 * in(i - 1, j)            &
     &                                  + 1.5 * in(i + 1, j)
                          else
                              out(i, j) = (n * rank + i) * in(i, j)
                          end if
30                    continue
20                continue
              end subroutine my_mv
              subroutine my_prec (n, in, out, mu)
                  use, intrinsic :: iso_c_binding, only: c_int, c_double
                  use mpi
                  ! size of the local matrix
                  integer (c_int), intent (in) :: n
                  ! number of right-hand sides
                  integer (c_int), intent (in) :: mu
                  ! input vector x
                  real (c_double), intent (in), dimension(n, mu)        &
     &                                                       :: in
                  ! output vector y = M^-1 x
                  real (c_double), intent (inout), dimension(n, mu)     &
     &                                                       :: out
                  integer :: i, j, rank
                  call MPI_Comm_rank(MPI_COMM_WORLD, rank, i)
                  ! simple diagonal scaling
                  do 20 j = 1, mu
                      do 30 i = 1, n
                          out(i, j) = in(i, j) / (n * rank + i)
30                    continue
20                continue
              end subroutine my_prec
      end module my_operator

      program main
          use, intrinsic :: iso_c_binding
          use hpddm
          use mpi
          use my_operator

          implicit none
          type (c_funptr) :: mv, precond
          real (c_double), pointer :: rhs(:, :), sol(:, :)
          integer (c_int) :: rank, ierr
          integer (c_int), dimension (1:2) :: dim

          call MPI_Init(ierr)
          call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

          mv = c_funloc(my_mv)
          precond = c_funloc(my_prec)
          if(rank == 0) then
              write(*, *) "Enter the size of the local matrices and then&
     & the number of right-hand sides:"
              read(*, *) dim(1), dim(2)
          end if
          call MPI_Bcast(dim, 2, MPI_INT, 0, MPI_COMM_WORLD, ierr)

          allocate(rhs(dim(1), dim(2)))
          allocate(sol(dim(1), dim(2)))

          ! config file
          ierr = HpddmParseConfig(C_CHAR_"hpddm_f90.cfg"//C_NULL_CHAR)
          if(rank /= 0) then
              call HpddmOptionRemove(C_CHAR_"verbosity"//C_NULL_CHAR)
          end if

          ! random RHSs and first guesses
          call random_number(rhs)
          call random_number(sol)

          ! HPDDM solve
          ierr = HpddmCustomOperatorSolve(dim(1), mv, precond, rhs, sol,&
     &                                    dim(2), MPI_COMM_WORLD)
          if(rank == 0) then
              write(*, *) "Number of iterations: ", ierr
          end if
          deallocate(rhs)
          deallocate(sol)
          call MPI_Finalize(ierr)
      end program main

File: examples/custom_operator.c
/*
   This file is part of HPDDM.

   Author(s): Pierre Jolivet <pierre.jolivet@enseeiht.fr>
        Date: 2020-05-07

   Copyright (C) 2020-     Centre National de la Recherche Scientifique

   HPDDM is free software: you can redistribute it and/or modify
   it under the terms of the GNU Lesser General Public License as published
   by the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   HPDDM is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with HPDDM.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <HPDDM.h>

struct HpddmCustomOperator {
    MPI_Comm comm;
    int n;
};

int mv(const HpddmCustomOperator* const op, const double* in, double* out, int mu) {
    int rank;
    MPI_Comm_rank(op->comm, &rank);
    for(int j = 0; j < mu; ++j)
        for(int i = 0; i < op->n; ++i) {
            out[op->n * j + i] = (op->n * rank + i + 2.0) * in[op->n * j + i];
            if(i > 0)
                out[op->n * j + i] -= 0.5 * in[op->n * j + i - 1];
            if(i < op->n - 1)
                out[op->n * j + i] -= 0.5 * in[op->n * j + i + 1];
        }
    return 0;
}
int apply(const HpddmCustomOperator* const op, const double* in, double* out, int mu) {
    int rank;
    MPI_Comm_rank(op->comm, &rank);
    for(int j = 0; j < mu; ++j)
        for(int i = 0; i < op->n; ++i)
            out[op->n * j + i] = in[op->n * j + i] / (op->n * rank + i + 2.0);
    return 0;
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    const HpddmOption* const opt = HpddmOptionGet();
    HpddmOptionParse(opt, argc, argv, 0);
    HpddmOptionParseString(opt, "-hpddm_verbosity 4");
    {
        char* val = "mu=<2>";
        char* desc = "Number of generated random right-hand sides.";
        HpddmOptionParseInt(opt, argc, argv, val, desc);
        val = "n=<100>";
        desc = "Size of the local matrices.";
        HpddmOptionParseInt(opt, argc, argv, val, desc);
    }
    int mu = HpddmOptionApp(opt, "mu");
    if(mu < 1)
        mu = 2;
    HpddmCustomOperator* op = (HpddmCustomOperator*)malloc(sizeof(HpddmCustomOperator));
    op->n = HpddmOptionApp(opt, "n");
    if(op->n < 1)
        op->n = 1;
    MPI_Comm_dup(MPI_COMM_WORLD, &op->comm);
    int rank, size, length;
    char name[MPI_MAX_PROCESSOR_NAME];
    MPI_Comm_rank(op->comm, &rank);
    if(rank != 0)
        HpddmOptionRemove(opt, "verbosity");
    MPI_Comm_size(op->comm, &size);
    MPI_Get_processor_name(name, &length);
    srand((unsigned)time(NULL)+ rank * size + length);
    double* in = malloc(sizeof(double) * mu * op->n);
    for(int i = 0; i < mu * op->n; ++i)
        in[i] = (rand() % 10000) / 100.0;
    double* out = (double*)calloc(mu * op->n, sizeof(double));
    HpddmCustomOperatorSolve(op, op->n, &mv, &apply, in, out, mu, &op->comm);
    MPI_Comm_free(&op->comm);
    free(out);
    free(in);
    free(op);
    MPI_Finalize();
    return 0;
}
